Update internal developer docs to use new configuration guidance (#5516)
alphagov/govuk-frontend: docs/contributing/coding-standards/component-options.md

Configuration options can be configured using Nunjucks, HTML or JavaScript, but only JavaScript will use the options.

> For more information, see our [guidance on building configurable components](https://frontend.design-system.service.gov.uk/building-configurable-components/).

Reviewer: **issue** The new documentation is a section of the existing 'Building your own JavaScript components' page.

```suggestion
For more information, see our [guidance on building configurable components](https://frontend.design-system.service.gov.uk/building-your-own-javascript-components/#building-configurable-components).
```

## Preparing the component's JavaScript

> If a component class needs configuration, it [can extend the `ConfigurableComponent` class](https://frontend.design-system.service.gov.uk/building-configurable-components/) that GOV.UK Frontend provides. This class accepts configuration from the following sources:

Reviewer: **nitpick** In the context of our internal configuration to describe how to create components for GOV.UK Frontend, I think extending `ConfigurableComponent` is more mandated than optional.

```suggestion
If a component class needs configuration, [extend the `ConfigurableComponent` class](https://frontend.design-system.service.gov.uk/building-configurable-components/) that GOV.UK Frontend provides. This class accepts configuration from the following sources:
```

- [defaults defined as a `defaults` static property](https://frontend.design-system.service.gov.uk/building-configurable-components/#setting-a-default-configuration-for-your-component) on the component’s class
- [configuration passed as second argument](https://frontend.design-system.service.gov.uk/building-configurable-components/#receiving-configuration-during-initialisation) to the component’s constructor or [`createAll`](https://frontend.design-system.service.gov.uk/building-your-own-javascript-components/#using-createall-with-your-components)
- [data attributes on the root element](https://frontend.design-system.service.gov.uk/building-configurable-components/#receiving-configuration-from-data-attributes) of the component

> These are the same sources as those used by GOV.UK Frontend components. Configuration options from these sources are merged into a single configuration object by the `ConfigurableComponent` class. Note that:

Reviewer: **issue** As we're inside the docs for GOV.UK Frontend, we can drop that part.

```suggestion
Configuration options from these sources are merged into a single configuration object by the `ConfigurableComponent` class. Note that:
```

- configuration options passed to the constructor will override any defaults
- data attribute values have highest precedence and will override any configuration set elsewhere

Your component will then be able to:

> [access the merged configuration](https://frontend.design-system.service.gov.uk/building-configurable-components/#accessing-the-component-s-configuration) using `this.config`
> use [the`Component` class](https://frontend.design-system.service.gov.uk/building-your-own-javascript-components/#using-the-component-class) features

Reviewer: I think this misses a list:
```suggestion
- [access the merged configuration](https://frontend.design-system.service.gov.uk/building-configurable-components/#accessing-the-component-s-configuration) using `this.config`
- use [the`Component` class](https://frontend.design-system.service.gov.uk/building-your-own-javascript-components/#using-the-component-class) features
```

export class Accordion extends GOVUKFrontendComponent {
  constructor($root, config = {}) {
    super($root)
- configuration options passed to the constructor will override any defaults
- data attribute values have highest precedence and will override any configuration set elsewhere

    this.config = mergeConfigs(
      Accordion.defaults,
      config,
      normaliseDataset(Accordion, this.$root.dataset)
    )
Your component will then be able to:

    // Check that the configuration provided is valid
    const errors = validateConfig(Accordion.schema, this.config)
    if (errors[0]) {
      throw new ConfigError(formatErrorMessage(CharacterCount, errors[0]))
    }
  }
}
```
[access the merged configuration](https://frontend.design-system.service.gov.uk/building-configurable-components/#accessing-the-component-s-configuration) using `this.config`
use [the`Component` class](https://frontend.design-system.service.gov.uk/building-your-own-javascript-components/#using-the-component-class) features

## Adding a Nunjucks parameter

> Most, but not all, components support adding arbitrary attributes and values through the `attributes` parameter. This method is also more verbose compared to having a dedicated Nunjucks parameter.

Reviewer: **suggestion** I think that bit is quite useful as it explains the two ways Nunjucks components can pass `data-*` attributes to the HTML for use by JavaScript components:
- via an `attributes` parameter allowing arbitrary `data-*` attributes 
- via specific Nunjucks options that get converted to `data-*` attributes

Author: Changed!

>   data-module="my-component"

Reviewer: **question** As these are internal docs, could we stick with the 'govuk' prefix? Maybe it's a matter of changing the introduction for the example, leading with something like:

'For example, this is how the Accordion does it for its `rememberExpanded` option:'

> We can now call the component's Nunjucks macro with our new `rememberExpanded` parameter:

Reviewer: **issue** The example following this statement still talks about the 'Accordion'. 

If we switch the introduction of the previous example to present the Accordion's template as an example, we might be able to do something similar here:

```suggestion
With the previous code in the Accordion's template, this is how a page may set a specific value for the `rememberExpanded` option:
```

### In Nunjucks

> Unlike the `data-*` attribute in HTML, there is no intrinsic link between the Nunjucks parameter name and the names used elsewhere. The Nunjucks parameter can therefore be named differently, if convenient.

Reviewer: **suggestion** Dropping 'and our use of `dataset`...' sounds good, but I think it needs replacing with something to explain we're talking about the link between the name of data attributes and JavaScript option names.
```suggestion
Unlike the `data-*` attribute in HTML and the name of JavaScript configuration options, there is no intrinsic link between the Nunjucks parameter name and the names used elsewhere. The Nunjucks parameter can therefore be named differently, if convenient.
```

Unlike the `data-*` attribute in HTML, there is no intrinsic link between the Nunjucks parameter name and the names used elsewhere. The Nunjucks parameter can therefore be named differently, if convenient.

A common case is specifying whether a parameter accepts HTML or only plain text. For example, if a configuration option's value is inserted into the page using `innerText`, you might want to name the Nunjucks parameter something like `sectionLabelText` to indicate that HTML will not be parsed if provided.

There is more guidance on naming Nunjucks parameters in [the Nunjucks API documentation](https://github.com/alphagov/govuk-frontend/blob/main/docs/contributing/coding-standards/nunjucks-api.md#naming-options).

## Namespacing configuration options

You can group configuration options in JavaScript and HTML together by using namespaces; period-separated strings that prefix the configuration option name. Namespaces follow the same formats as other option names, being camelCase in JavaScript and kebab-case in HTML.

> These are most commonly used for translation strings, which are usually namespaced under `i18n` (short for 'internationalisation').

Reviewer: **suggestion** Can we keep that as an explanation for the use of nested objects as it's the main one we have in `govuk-frontend`? Maybe to form a paragraph with "When passing configuration into your component's constructor...", moving the explanation on how to use data attributes to a second paragraph?

> When passing configuration into your component's constructor, you can use nested objects to group options. This isn't possible when using data attributes, which only accept strings. Instead, you should follow our [naming conventions for data attributes](https://frontend.design-system.service.gov.uk/building-configurable-components/#receiving-configuration-from-data-attributes).

Reviewer: **suggestion** The "This isn't possible" reads as if you won't be able to set nested options through data attributes. Maybe we can spin this more positively with:

```suggestion
When passing configuration into your component's constructor, you can use nested objects to group options. 

Data attributes only accept strings, but our  [naming conventions for data attributes](https://frontend.design-system.service.gov.uk/building-configurable-components/#receiving-configuration-from-data-attributes) allow you to set options in nested objects using a dot `.` separator for each level of nesting. For example `data-i18n.showLabel="Show"` will set the same option as the following constructor argument:

```json
{
  i18n: {
     showLabel: 'Show'
  }
}
```
